WalletWasabi/WabiSabi/Client/CoinJoinClient.cs

```diff
@@ -48,6 +48,7 @@ public class CoinJoinClient
 		public Kitchen Kitchen { get; }
 		public KeyManager Keymanager { get; }
 		private RoundStateUpdater RoundStatusUpdater { get; }
+		private Random Random { get; } = new Random();
```

Comment:
Can't we use `SecureRandom` that is already in this class? Or is it too much in this case?

Reply:
Is there something to gain if we were be able to set precise delays in tests?

Reply:
There is nothing to gain and using a cryptographic secure random generator for something like this sound like a bad idea. When I had to track the source of entropy for the wallet generation process i found that dotnet uses pbkdf2 on top of the CSPRNG api so, it is a lot of computation for something really trivial.

Reply:
wait what? if an adverserial coordinator can predict the RNG seed for the initial value, which IIRC is seeded by time, it can predict all of the delays... we definitely want to at least seed `Random` with `SecureRandom`, but I would prefer generating numbers with it as well, it's only a few numbers that we need so the computational overhead is minimal.

the reason wallet generation is slow is precisely to do with the keystretching, i.e. it's slow by design, that doesn't have much to do with `SecureRandom` itself

Reply:
There are two different points here. Let start with the easiest one: dotnet BLC crypto api uses pdkdf2 on top of the native windows crypto api. That's a fact, i can prove it. That's a slow thing wrapping another slow thing, that's what our `SecureRandom` uses. But given discussions about performance is what idiots always do, i will never mention this again.

I didn't considered the fact that `Random` is deterministic. Clearly we have to change it. Let me see what's the more convenient way to proceed.

Reply:
OK i thought you were referring to the use of scrypt in nbitcoin which was slowing down the tests.


> There are two different points here. Let start with the easiest one: dotnet BLC crypto api uses pdkdf2 on top of the native windows crypto api. That's a fact, i can prove it.

Please do, this seems *EXTREMELY* poor and dangerous design and a problem to be aware of, but a bit of searching yielded no result.

It's not a security problem if applied to sufficient starting entropy, only a performance issue, but the P in PBKDF2 stands for "password", i.e. it's specifically designed for keystretching, the only purpose of that function is to slow things down when there is no access to a good source of entropy, which makes no sense for a CSPRNG which is meant to be seeded with sufficient entropy. If the use of PBKDF2 was appropriate, this implies that RandomNumberGenerator is not a CSPRNG at all!

Either way, we should get to the bottom of this

> I didn't considered the fact that `Random` is deterministic. Clearly we have to change it. Let me see what's the more convenient way to proceed.

The big hassle is getting bounded ints out of a byte oriented API, but I think we should just do the safe thing, rely on `SecureRandom`, and fix the performance issue if it's actually a problem (I doubt it because we generate a lot more random data in the ZK proofs and that never seemed to be a performance bottleneck)

Reply:
Sorry, it is bcrypt. Here you have: https://github.com/bitcoin-dot-org/Bitcoin.org/pull/2788#issuecomment-454482200

Reply:
FWIW what I'm seeing here doesn't seem to line up: https://github.com/dotnet/runtime/tree/main/src/libraries/System.Security.Cryptography.Algorithms/src/Internal/Cryptography

Specifically, on Windows bcrypt is used:  https://github.com/dotnet/runtime/blob/b3c592f1c9b5df7f572a2f96582412655695f5db/src/libraries/Common/src/Interop/Windows/BCrypt/Interop.BCryptGenRandom.cs#L14

Which goes to:

https://github.com/dotnet/runtime/blob/main/src/libraries/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/RandomNumberGeneratorImplementation.Windows.cs#L14

which is an FFI to the windows API: https://github.com/dotnet/runtime/blob/b3c592f1c9b5df7f572a2f96582412655695f5db/src/libraries/Common/src/Interop/Windows/BCrypt/Interop.BCryptGenRandom.cs#L14

which suggests otherwise https://docs.microsoft.com/en-us/windows/win32/api/bcrypt/nf-bcrypt-bcryptgenrandom

because with the supplied parameters it always uses the system preferred random number generator